File: src/generators/templates/python/proxy.py
# === Sandbox Executor Python Proxy (Stdin Mode) ===
import sys, json, traceback, importlib.util

# Result markers
START_MARKER = "__SANDBOX_RESULT_START__"
END_MARKER = "__SANDBOX_RESULT_END__"

# Module cache for potential persistent execution
_USER_MODULE_CACHE = {}

def __sandbox_serialize__(obj):
    if hasattr(obj, '__dict__'):
        return {k: __sandbox_serialize__(v) for k, v in obj.__dict__.items() if not k.startswith('_')}
    if hasattr(obj, 'to_dict'):
        return obj.to_dict()
    if isinstance(obj, (set, frozenset)):
        return list(obj)
    if isinstance(obj, bytes):
        return obj.decode('utf-8', errors='replace')
    if hasattr(obj, '__iter__') and not isinstance(obj, (str, dict, list, tuple)):
        return list(obj)
    return obj

def execute_request(request):
    try:
        file_path = request.get('filePath', '/workspace/user_code.py')
        func_name = request['functionName']
        args = request.get('args', [])
        kwargs = request.get('kwargs', {})

        if file_path not in _USER_MODULE_CACHE:
            spec = importlib.util.spec_from_file_location("user_code", file_path)
            module = importlib.util.module_from_spec(spec)
            spec.loader.exec_module(module)
            _USER_MODULE_CACHE[file_path] = module
        
        module = _USER_MODULE_CACHE[file_path]
        func = getattr(module, func_name)
        
        result = func(*args, **kwargs)
        return {
            "success": True,
            "result": __sandbox_serialize__(result)
        }
    except Exception as e:
        return {
            "success": False,
            "error": {
                "message": str(e),
                "type": type(e).__name__,
                "stack": traceback.format_exc()
            }
        }

def main():
    try:
        # Read protocol header: [Mode(1b)][Length(4b)]
        header = sys.stdin.buffer.read(5)
        if len(header) < 5:
            return
        
        mode = chr(header[0])
        length = int.from_bytes(header[1:5], 'big')
        
        if mode == 'A': # Atomic mode
            payload = sys.stdin.buffer.read(length)
            if not payload:
                raise ValueError("Empty payload")
            request = json.loads(payload.decode('utf-8'))
            output = execute_request(request)
            
            print(START_MARKER)
            print(json.dumps(output, ensure_ascii=False))
            print(END_MARKER)
            sys.stdout.flush()
    except Exception as e:
        output = {
            "success": False,
            "error": {
                "message": "Fatal: " + str(e),
                "type": type(e).__name__,
                "stack": traceback.format_exc()
            }
        }
        print(START_MARKER)
        print(json.dumps(output, ensure_ascii=False))
        print(END_MARKER)

if __name__ == "__main__":
    main()

File: src/generators/templates/ruby/proxy.rb
# === Sandbox Executor Ruby Proxy ===
require 'json'

START_MARKER = "__SANDBOX_RESULT_START__"
END_MARKER = "__SANDBOX_RESULT_END__"

def execute_request(request)
  begin
    file_path = request['filePath'] || '/workspace/user_code.rb'
    func_name = request['functionName']
    args = request['args'] || []
    kwargs = request['kwargs'] || {}

    require_relative file_path.sub(/\.rb$/, '')

    # Ruby 2.7+ handles kwargs explicitly
    if kwargs && !kwargs.empty?
      symbol_kwargs = kwargs.transform_keys(&:to_sym)
      result = send(func_name, *args, **symbol_kwargs)
    else
      result = send(func_name, *args)
    end

    {
      "success" => true,
      "result" => result
    }
  rescue => e
    {
      "success" => false,
      "error" => {
        "message" => e.message,
        "type" => e.class.name,
        "stack" => e.backtrace.join("
")
      }
    }
  end
end

begin
  header = STDIN.read(5)
  if header && header.length == 5
    mode = header[0]
    length = header[1..4].unpack1('N') # Big-endian 32-bit unsigned

    if mode == 'A'
      payload = STDIN.read(length)
      if payload
        request = JSON.parse(payload)
        output = execute_request(request)
        puts START_MARKER
        puts JSON.generate(output)
        puts END_MARKER
      else
        raise "Empty payload"
      end
    end
  else
    raise "Header too short or stdin empty"
  end
rescue => e
  output = {
    "success" => false,
    "error" => {
      "message" => "Fatal: " + e.message,
      "type" => e.class.name,
      "stack" => e.backtrace&.join("
")
    }
  }
  puts START_MARKER
  puts JSON.generate(output)
  puts END_MARKER
end
